File: mobile_karyawan/lib/data/services/attendance_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import '../../core/constants/app_constants.dart';
import '../../models/attendance_model.dart';
import 'auth_service.dart';

class AttendanceService {
  final String _baseUrl = AppConstants.baseUrl;
  final AuthService _authService = AuthService();

  Future<Map<String, String>> _getHeaders() async {
    final token = await _authService.getToken();
    return {
      'Content-Type': 'application/json',
      'Authorization': 'Bearer $token',
    };
  }

  Future<List<Attendance>> getHistory() async {
    final response = await http.get(
      Uri.parse('$_baseUrl/attendance/history'),
      headers: await _getHeaders(),
    );

    if (response.statusCode == 200) {
      List data = jsonDecode(response.body);
      return data.map((item) => Attendance.fromJson(item)).toList();
    }
    return [];
  }

  Future<Map<String, dynamic>> checkIn(Map<String, dynamic> data) async {
    final response = await http.post(
      Uri.parse('$_baseUrl/attendance/check-in'),
      headers: await _getHeaders(),
      body: jsonEncode(data),
    );
    final body = jsonDecode(response.body);
    if (response.statusCode != 200 && response.statusCode != 201) {
      throw Exception(body['message'] ?? 'Check-in gagal');
    }
    return body;
  }

  Future<Map<String, dynamic>> checkOut(int attendanceId, Map<String, dynamic> data) async {
    final response = await http.post(
      Uri.parse('$_baseUrl/attendance/check-out'),
      headers: await _getHeaders(),
      body: jsonEncode({
        'attendance_id': attendanceId,
        ...data,
      }),
    );
    return jsonDecode(response.body);
  }

  /// HRD: daily summary — semua karyawan berjadwal pada tanggal tertentu
  /// [date] format: 'YYYY-MM-DD', null = hari ini
  Future<Map<String, dynamic>?> getDailySummary({String? date}) async {
    final uri = Uri.parse(
      date != null
          ? '$_baseUrl/attendance/daily-summary?date=$date'
          : '$_baseUrl/attendance/daily-summary',
    );
    final response = await http.get(uri, headers: await _getHeaders());
    if (response.statusCode == 200) {
      return jsonDecode(response.body) as Map<String, dynamic>;
    }
    return null;
  }
}

File: web-hrd/lib/services/attendanceService.js
const API_URL = process.env.NEXT_PUBLIC_API_URL;

const getHeaders = () => {
  const token = localStorage.getItem("token");
  return {
    "Content-Type": "application/json",
    Authorization: `Bearer ${token}`,
  };
};

export const attendanceService = {
  getHistory: async (date) => {
    let url = `${API_URL}/attendance/history`;
    if (date) url += `?date=${date}`;
    const response = await fetch(url, { headers: getHeaders() });
    return response.json();
  },
  checkIn: async (data) => {
    const response = await fetch(`${API_URL}/attendance/check-in`, {
      method: "POST",
      headers: getHeaders(),
      body: JSON.stringify(data),
    });
    return response.json();
  },
  checkOut: async (data) => {
    const response = await fetch(`${API_URL}/attendance/check-out`, {
      method: "POST",
      headers: getHeaders(),
      body: JSON.stringify(data),
    });
    return response.json();
  },
  getMonthlySummary: async ({ month, year, userId } = {}) => {
    const params = new URLSearchParams();
    if (month) params.set("month", month);
    if (year)  params.set("year", year);
    if (userId) params.set("user_id", userId);
    const response = await fetch(`${API_URL}/attendance/monthly-summary?${params}`, {
      headers: getHeaders(),
    });
    return response.json();
  },
};
